File: conversions/color-convert/elide-color-convert.py
#!/usr/bin/env python3
"""
Python Integration Example for elide-color-convert

This demonstrates calling the TypeScript color conversion implementation from Python
using Elide's polyglot capabilities.

Benefits:
- One color conversion implementation shared across TypeScript and Python
- Consistent color handling across design systems
- No Python colorsys complexity
- Supports RGB, HSL, HSV, HEX conversions
"""

# NOTE: Exact syntax depends on Elide's Python polyglot API (currently alpha)
# This is a CONCEPTUAL example - adjust when Python support is ready

# Assuming Elide provides something like:
# from elide import require
# color = require('./elide-color-convert.ts')

print("=== Python Calling TypeScript Color Convert ===\n")

# Example 1: RGB to HEX conversion
# hex_color = color.rgbToHex([255, 0, 0])
# print(f"RGB(255, 0, 0) -> {hex_color}")  # #ff0000
# print()

# Example 2: HEX to RGB conversion
# rgb = color.hexToRgb("#ff0000")
# print(f"#ff0000 -> RGB{rgb}")  # [255, 0, 0]
# print()

# Example 3: Image processing pipeline
# def process_image_colors(pixels):
#     """Process image pixels with color transformations"""
#     processed = []
#     for rgb in pixels:
#         # Convert to HSL
#         hsl = color.rgbToHsl(rgb)
#
#         # Adjust lightness
#         hsl[2] = min(100, hsl[2] + 20)  # Lighten by 20%
#
#         # Convert back to RGB
#         new_rgb = color.hslToRgb(hsl)
#         processed.append(new_rgb)
#
#     return processed
#
# sample_pixels = [[100, 150, 200], [50, 75, 100]]
# result = process_image_colors(sample_pixels)
# print("Processed image colors:", result)
# print()

# Example 4: Design system color generation
# def generate_color_palette(base_color_hex):
#     """Generate a color palette from base color"""
#     rgb = color.hexToRgb(base_color_hex)
#
#     return {
#         'base': base_color_hex,
#         'light': color.rgbToHex(color.lighten(rgb, 20)),
#         'dark': color.rgbToHex(color.darken(rgb, 20)),
#         'saturated': color.rgbToHex(color.saturate(rgb, 30)),
#         'complement': color.rgbToHex(color.complement(rgb))
#     }
#
# palette = generate_color_palette("#3b82f6")
# print("Generated palette:")
# for variant, hex_val in palette.items():
#     print(f"  {variant}: {hex_val}")
# print()

# Example 5: Data visualization
# def color_for_value(value, min_val, max_val):
#     """Generate color based on value (heatmap)"""
#     # Normalize value to 0-1
#     normalized = (value - min_val) / (max_val - min_val)
#
#     # Generate HSL color (0° = red, 120° = green)
#     hue = int(normalized * 120)
#     hsl = [hue, 100, 50]
#
#     # Convert to RGB and HEX
#     rgb = color.hslToRgb(hsl)
#     return color.rgbToHex(rgb)
#
# values = [10, 50, 90]
# print("Heatmap colors:")
# for val in values:
#     hex_color = color_for_value(val, 0, 100)
#     print(f"  Value {val}: {hex_color}")
# print()

print("Real-world use case:")
print("- Python data visualization needs consistent color conversions")
print("- Uses same TypeScript implementation as design system")
print("- Guarantees identical colors across all services")
print("- No need to install Python color library")
print()

print("Example: Design System")
print("┌─────────────────────────────────────┐")
print("│   Elide Color Convert (TypeScript) │")
print("│   elide-color-convert.ts           │")
print("└─────────────────────────────────────┘")
print("         ↓                   ↓")
print("    ┌────────┐          ┌────────┐")
print("    │ React  │          │ Python │")
print("    │   UI   │          │ Charts │")
print("    └────────┘          └────────┘")
print("         ↓                   ↓")
print("    Same color values everywhere!")
print()

print("Problem Solved:")
print("Before: Different color libs = inconsistent colors across UI and charts")
print("After: One Elide implementation = identical color output")
print()

print("When Elide Python API is ready, usage will be:")
print("  from elide import require")
print("  color = require('./elide-color-convert.ts')")
print("  hex_color = color.rgbToHex([255, 0, 0])")
print("  print(hex_color)  # #ff0000")

File: converted/utilities/color-convert/elide-color-convert.rb
#!/usr/bin/env ruby

# Ruby Integration Example for elide-color-convert
#
# This demonstrates calling the TypeScript color conversion implementation from Ruby
# using Elide's polyglot capabilities.
#
# Benefits:
# - One color conversion implementation shared across TypeScript and Ruby
# - Consistent color handling across design systems
# - No Ruby color gem complexity
# - Perfect for Rails UI theming

puts "=== Ruby Calling TypeScript Color Convert ===\n"

# NOTE: Exact syntax depends on Elide's Ruby polyglot API (currently alpha)
# This is a CONCEPTUAL example - adjust when Ruby support is ready

# Assuming Elide provides something like:
# color = Elide.require('./elide-color-convert.ts')

# Example 1: Rails UI theming
# class Theme < ApplicationRecord
#   def generate_variants
#     color = Elide.require('./elide-color-convert.ts')
#     base_rgb = color.hexToRgb(primary_color)
#
#     {
#       primary: primary_color,
#       light: color.rgbToHex(color.lighten(base_rgb, 20)),
#       dark: color.rgbToHex(color.darken(base_rgb, 20)),
#       accent: color.rgbToHex(color.complement(base_rgb))
#     }
#   end
# end

# Example 2: CSS generation
# def generate_css_variables(theme_colors)
#   color = Elide.require('./elide-color-convert.ts')
#   css = ":root {\n"
#
#   theme_colors.each do |name, hex|
#     rgb = color.hexToRgb(hex)
#     css += "  --color-#{name}: #{hex};\n"
#     css += "  --color-#{name}-rgb: #{rgb.join(', ')};\n"
#   end
#
#   css += "}"
#   css
# end

# Example 3: Image processing
# class ImageProcessor
#   def adjust_brightness(image_data, amount)
#     color = Elide.require('./elide-color-convert.ts')
#
#     image_data.map do |pixel|
#       rgb = pixel[:rgb]
#       adjusted = amount > 0 ?
#         color.lighten(rgb, amount) :
#         color.darken(rgb, -amount)
#       { rgb: adjusted, hex: color.rgbToHex(adjusted) }
#     end
#   end
# end

# Example 4: Color palette generation
# def create_palette(base_color)
#   color = Elide.require('./elide-color-convert.ts')
#   base_rgb = color.hexToRgb(base_color)
#   hsl = color.rgbToHsl(base_rgb)
#
#   # Generate complementary colors
#   palette = []
#   [0, 30, 60, 90].each do |hue_shift|
#     new_hsl = [hsl[0] + hue_shift, hsl[1], hsl[2]]
#     new_rgb = color.hslToRgb(new_hsl)
#     palette << color.rgbToHex(new_rgb)
#   end
#
#   palette
# end

puts "Real-world use case:"
puts "- Rails app needs consistent color theming"
puts "- Uses same TypeScript implementation as React frontend"
puts "- Guarantees identical color values across stack"
puts "- No gem installation needed"
puts ""

puts "Example: Design System"
puts "┌─────────────────────────────────────┐"
puts "│   Elide Color Convert (TypeScript) │"
puts "│   elide-color-convert.ts           │"
puts "└─────────────────────────────────────┘"
puts "         ↓           ↓           ↓"
puts "    ┌────────┐  ┌────────┐  ┌────────┐"
puts "    │ React  │  │  Rails │  │Sidekiq │"
puts "    │   UI   │  │Backend │  │ Worker │"
puts "    └────────┘  └────────┘  └────────┘"
puts "         ↓           ↓           ↓"
puts "    Same color system everywhere!"
puts ""

puts "Problem Solved:"
puts "Before: Different color libs = UI and backend generate different colors"
puts "After: One Elide implementation = 100% color consistency"
puts ""

puts "Rails Integration (when ready):"
puts "  # config/initializers/color.rb"
puts "  COLOR = Elide.require('./elide-color-convert.ts')"
puts "  "
puts "  # app/helpers/theme_helper.rb"
puts "  module ThemeHelper"
puts "    def lighten_color(hex, amount)"
puts "      rgb = COLOR.hexToRgb(hex)"
puts "      COLOR.rgbToHex(COLOR.lighten(rgb, amount))"
puts "    end"
puts "  end"
